packages/vite/src/node/config.ts

```diff
@@ -1232,11 +1224,48 @@ export function isDepsOptimizerEnabled(
   config: ResolvedConfig,
   ssr: boolean,
 ): boolean {
-  const { command } = config
-  const { disabled } = getDepOptimizationConfig(config, ssr)
-  return !(
-    disabled === true ||
-    (command === 'build' && disabled === 'build') ||
-    (command === 'serve' && disabled === 'dev')
-  )
+  const optimizeDeps = getDepOptimizationConfig(config, ssr)
+  return !(optimizeDeps.noDiscovery && !optimizeDeps.include?.length)
+}
+
+function optimizeDepsDisabledBackwardCompatibility(
+  resolved: ResolvedConfig,
+  optimizeDeps: DepOptimizationConfig,
+  optimizeDepsPath: string = '',
+) {
+  const optimizeDepsDisabled = optimizeDeps.disabled
+  if (optimizeDepsDisabled !== undefined) {
+    if (optimizeDepsDisabled === true || optimizeDepsDisabled === 'dev') {
+      const commonjsOptionsInclude = resolved.build?.commonjsOptions?.include
+      const commonjsPluginDisabled =
+        Array.isArray(commonjsOptionsInclude) &&
+        commonjsOptionsInclude.length === 0
+      optimizeDeps.noDiscovery = true
+      optimizeDeps.include = undefined
+      if (commonjsPluginDisabled) {
+        resolved.build.commonjsOptions.include = undefined
+      }
+      resolved.logger.warn(
+        colors.yellow(`(!) Experimental ${optimizeDepsPath}optimizeDeps.disabled and deps pre-bundling during build were removed in Vite 5.1.
+    To disable the deps optimizer, set ${optimizeDepsPath}optimizeDeps.noDiscovery to true and ${optimizeDepsPath}optimizeDeps.include as undefined or empty. 
+    Please remove ${optimizeDepsPath}optimizeDeps.disabled from your config.
+    ${
+      commonjsPluginDisabled
+        ? 'Empty config.build.commonjsOptions.include will be ignored to support CJS during build. This config should also be removed.'
+        : ''
```

Comment:
For the warning strings, is it intentional that they have the indent and line breaks here? I think it might be worth removing the indents, or doing something like

```
`some long string\n` + `other string`
```

to clean it up a bit, and so it displays nicely for smaller terminal sizes too.

Reply:
The indents are intentional. Would this look better to you?
```js
      resolved.logger.warn(
        colors.yellow(
          `(!) Experimental ${optimizeDepsPath}optimizeDeps.disabled and deps pre-bundling during build were removed in Vite 5.1.\n` +
            `    To disable the deps optimizer, set ${optimizeDepsPath}optimizeDeps.noDiscovery to true and ${optimizeDepsPath}optimizeDeps.include as undefined or empty.\n` +
            `    Please remove ${optimizeDepsPath}optimizeDeps.disabled from your config.\n` +
            commonjsPluginDisabled
            ? '    Empty config.build.commonjsOptions.include will be ignored to support CJS during build. This config should also be removed.\n'
            : '',
        ),
      )
```

Reply:
Hmm if you want to keep the indents, then I think the original code is fine. I personally don't indent and new line messages since it always assumes a certain terminal width 😅 

I'm fine if you want to resolve this conversation as is, not a big blocker.

Reply:
Ok, let's keep it as it should only appear for very few folks. My idea was to give them a clear warning and the extra lines were helping in this case.